wemake_python_styleguide/visitors/ast/builtins.py

```diff
@@ -125,6 +124,128 @@ def _check_modulo_patterns(
+        ast.Subscript,
+        ast.Attribute,
+    )
+
+    def visit_JoinedStr(self, node: ast.JoinedStr) -> None:
+        """
+        Forbids use of ``f`` strings and too complex ``f`` strings.
+
+        Raises:
+            FormattedStringViolation
+            TooComplexFormattedStringViolation
+
+        """
+        self.add_violation(consistency.FormattedStringViolation(node))
+        self.generic_visit(node)
+        self._check_complex_formatted_string(node)
+
+    def _check_complex_formatted_string(self, node: ast.JoinedStr) -> None:
+        # Whitelists all simple uses of f strings
+        # Checks if list, dict, function call with no parameters or variable
+        has_formatted_components = any(
+            isinstance(comp, ast.FormattedValue) for comp in node.values
+        )
+        if not has_formatted_components:
+            # If no formatted values
+            self.add_violation(
+                complexity.TooComplexFormattedStringViolation(node),
+            )
+            return
+
+        for string_component in node.values:
+            if isinstance(string_component, ast.FormattedValue):
+                # Test if possible chaining is invalid
+                format_value = string_component.value
+                if self._is_valid_formatted_value(format_value):
+                    continue
+                # Everything else is too complex
+                self.add_violation(
+                    complexity.TooComplexFormattedStringViolation(node),
+                )
+                break
+
+    def _is_valid_formatted_value(self, format_value: ast.AST) -> bool:
+        if isinstance(format_value, (ast.Call, ast.Subscript, ast.Attribute)):
+            if not self._is_valid_chaining(format_value):
+                return False
+        return self._is_valid_final_value(format_value)
+
+    def _is_valid_final_value(self, format_value: ast.AST) -> bool:
+        # Variable lookup is okay and a single attribute is okay
+        if isinstance(format_value, (ast.Name, ast.Attribute)):
+            return True
+        # Function call with empty arguments is okay
+        elif isinstance(format_value, ast.Call) and not format_value.args:
+            return True
+        # Named lookup, Index lookup & Dict key is okay
+        elif isinstance(format_value, ast.Subscript):
+            if isinstance(format_value.slice, ast.Index):
+                return isinstance(
+                    format_value.slice.value,
+                    self._valid_format_index,
+                )
+        return False
+
+    def _is_valid_chaining(self, format_value: AnyChainable) -> bool:
+        parts = iter(attributes.parts(format_value))
+        chained_parts: List[ast.AST] = []
+
+        iteration = 0
+        try:
+            # Should throw on at latest the fifth attempt
+            while iteration < 5:
+                chained_parts.append(next(parts))
+                iteration += 1
+        except StopIteration:
```

Comment:
We try not to use exceptions, they are hard to test. Can you please rewrite this part?